client/components/LearnerViewDirectiveSpec.js

```diff
@@ -0,0 +1,201 @@
+    it('should navigate to given question', function() {
+      for (var i = 0; i < $scope.questionIds.length; i++) {
+        $scope.navigateToQuestion(i);
+        expect($scope.currentQuestionIndex).toBe(i);
+      }
+    });
+  });
+
+  describe("activateAutosaving", function() {
+    var $interval;
+    var $timeout;
+    var CodeStorageService;
+    beforeEach(inject(function(_$interval_, _$timeout_, _CodeStorageService_) {
+      $interval = _$interval_;
+      $timeout = _$timeout_;
+      CodeStorageService = _CodeStorageService_;
+      localStorage.clear();
+    }));
+
+    // Autosave text is using $timeout, thus,
+    // we need to flush both $interval and $timeout.
+    var timeFlies = function(timeToFly) {
+      $interval.flush(timeToFly);
+      $timeout.flush(timeToFly);
+    };
+
+    it("should activate autosaving for only once", function() {
+      for (var i = 0; i < $scope.questionIds.length; i++) {
+        if (i !== 0) {
+          // Default question index is 0, we don't want to
+          // navigate to question 0 again because navigateToQuestion
+          // will store code for current question, thus, fail the test
+          $scope.navigateToQuestion(i);
+        }
+        expect(!$scope.autosaveOn).toBe(true);
+        var questionId = $scope.questionIds[i];
+        // There should be no code stored in local storage
+        // before autosave is triggerred.
+        expect(CodeStorageService.loadStoredCode(
+          questionId, LANGUAGE)).toEqual(null);
+        $scope.activateAutosaving();
+        expect($scope.autosaveOn).toBe(true);
+        // Flush 4900 milliseconds -- time: 4.9s
+        expect($scope.autosaveTextIsDisplayed).toBe(false);
+        timeFlies(AUTOSAVE_MILLISECONDS - DELTA_MILLISECONDS);
+        expect($scope.autosaveOn).toBe(true);
+        expect(CodeStorageService.loadStoredCode(
+          questionId, LANGUAGE)).toEqual(null);
+        // Flush 200 milliseconds -- time: 5.1s
+        timeFlies(2 * DELTA_MILLISECONDS);
+        var question = QuestionDataService.getQuestion(questionId);
+        var starterCode = question.getStarterCode(LANGUAGE);
+        expect(CodeStorageService.loadStoredCode(
+          questionId, LANGUAGE)).toEqual(starterCode);
+        expect($scope.autosaveOn).toBe(true);
+        expect($scope.autosaveTextIsDisplayed).toBe(true);
+        // Flush 1000 milliseconds -- time: 6.1s
+        timeFlies(SECONDS_TO_MILLISECONDS);
+        expect($scope.autosaveTextIsDisplayed).toBe(false);
+        // Flush 4000 milliseconds -- time: 10.1s
+        timeFlies(4 * SECONDS_TO_MILLISECONDS);
+        expect($scope.autosaveTextIsDisplayed).toBe(false);
+        expect($scope.autosaveOn).toBe(false);
+        expect(CodeStorageService.loadStoredCode(
+          questionId, LANGUAGE)).toEqual(starterCode);
+      }
+    });
+
+    it("should store the latest code into localStorage", function() {
+      for (var i = 0; i < $scope.questionIds.length; i++) {
```

Comment:
Honestly, I don't think I will need the first part where it's similar to the previous test. I put it there just to be safe I guess. What I really want to test here is, with continuous change, the autosave is working properly. I want to test that by checking the latest code is stored in localStorage. Thus, the initial status doesn't really matter here. Do you think I should just get rid of the similar part? @eyurko